src/Form/PolicyForm.php

```diff
@@ -0,0 +1,224 @@
+    $entity = $this->entity;
+
+    $form['#title'] = $this->t('Edit policy');
+
+    $options = [];
+    $current_role = $entity->getRole();
+    $existing_policies = $this->policyConfigEntityStorage->loadMultiple();
+    foreach ($this->roleStorage->loadMultiple() as $role) {
+      $id = $role->id();
+      if ($id === AccountInterface::ANONYMOUS_ROLE || (array_key_exists($id, $existing_policies) && $current_role !== $id)) {
+        continue;
+      }
+
+      $options[$id] = $role->label();
+    }
+
+    $form['role'] = [
+      '#type' => 'select',
+      '#options' => $options,
+      '#title' => $this->t('User role'),
+      '#description' => $this->t('Select a role for which this policy should be applied.<br>When the user is registering only the <em>Authenticated user</em> role policy will be used.'),
+      '#default_value' => $current_role,
+      '#required' => TRUE,
+      '#disabled' => !$entity->isNew(),
+    ];
+
+    $form['minimumRequiredConstraints'] = [
+      '#type' => 'number',
+      '#title' => $this->t('Minimum required constraints'),
+      '#min' => 1,
+      '#required' => TRUE,
+      '#default_value' => $entity->getMinimumRequiredConstraints(),
+    ];
+
+    $expire_days = $entity->getExpireDays();
+    $form['expire_password'] = [
+      '#type' => 'checkbox',
+      '#title' => $this->t('Expire password'),
+      '#default_value' => $expire_days > 0,
+    ];
+
+    $form['expire_days'] = [
+      '#type' => 'number',
+      '#title' => $this->t('Expire password after'),
+      '#field_suffix' => $this->t('day(s).'),
+      '#default_value' => $expire_days ?: 1,
+      '#min' => 1,
+      '#states' => [
+        'visible' => [
+          ':input[name="expire_password"]' => ['checked' => TRUE],
+        ],
+      ],
+    ];
+
+    $form['expire_warn_before_days'] = [
+      '#type' => 'number',
+      '#title' => $this->t('Show warning'),
+      '#description' => $this->t('Show a warning message before the password would expire.<br>If set to 0 it will not show any message.'),
+      '#field_suffix' => $this->t('day(s) before the password expires.'),
+      '#default_value' => $entity->getExpireWarnDays(),
+      '#min' => 0,
+      '#states' => [
+        'visible' => [
+          ':input[name="expire_password"]' => ['checked' => TRUE],
+        ],
+      ],
+    ];
+
+    $form['expiryWarningMessage'] = [
+      '#type' => 'textarea',
+      '#title' => $this->t('Expiry warning message'),
+      '#description' => $this->t("This message will be presented for the user if the user's password is about to expire.<br>Useable placeholders: <ul><li>@url: URL for the user edit page</li><li>@date_time: the date and time when the password will expire</li></ul>."),
+      '#default_value' => $entity->getExpiryWarningMessage() ?? 'Your password will expire on @date_time, please <a href="@url">change your password</a> before it expires to prevent any potential data loss.',
```

Comment:
Shouldn't `$entity->getExpiryWarningMessage()` return this message by default instead of a `null`? Based on this usage all implementation of `PolicyInterface` should return a (non-empty) string when this method gets called.

Reply:
It is not mandatory to provide the warning message, and I didn't wanted to store the default message on the entity itself, only where it is being set on the UI.

Reply:
It is not mandatory to be set, but it seems you are using it as it would be. Just check this code of yours:

```php
<?php
  public function passwordChangeNotification(GetResponseEvent $event) {
    if ($event->isMasterRequest() && Url::fromRoute('password_enhancements.password_change')->toString() !== $event->getRequest()->getPathInfo()) {
      $policy = $this->policyStorage->loadByRoleAndPriority($this->account->getRoles());
      if ($policy !== NULL && $this->passwordChecker->showWarningMessage($policy)) {
        $user = $this->userStorage->load($this->account->id());
        $this->messenger->addWarning(new FormattableMarkup($policy->getExpiryWarningMessage(), [
          '@date_time' => $this->dateFormatter->format($user->password_enhancements_password_changed_date->value + $policy->getExpireSeconds(), 'password_enhancements_date_format'),
          '@url' => Url::fromRoute('entity.user.edit_form', ['user' => $user->id()])->toString(),
        ]));
      }
    }
  }
```

I do not see that would ever check whether the `getExpiryWarningMessage()` returns null or not.

And probably it makes sense that if the user configured to show a warning message then the system should display a default message unless the user customized it.

Reply:
(Ok, so it can not be customized whether a warning should be displayed or not, but my argument is still valid.)

Reply:
No, I'm not checking whether it returns null or not, to show the message it is determined by the showWarningMessage() method.

Reply:
Follow up task #4